### PR TITLE
Fix resource stream for collectible assemblies

### PR DESCRIPTION
**Pull request by janvorli**
The GetManifestResourceStream was returning a stream that didn't keep
the underlying assembly alive. For collectible assemblies, that means
that an assembly could be collected and the stream still kept,
potentially reading garbage.

The fix is to create a new stream type that stores a reference to the
underlying assembly, thus ensuring the proper lifetime management.

I have also added a regression test that checks that the stream keeps the underlying assembly alive and it also checks that once the stream reference is gone, the assembly is collected.

### CODE REVIEW

**Comment by janvorli** [comment]
cc: @marek-safar

File: src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs [hunk 1]
```diff
@@ -29,6 +29,16 @@ internal class RuntimeAssembly : Assembly
 
         #endregion
 
+        private class ManifestResourceStream : UnmanagedMemoryStream
+        {
+            private RuntimeAssembly _manifestAssembly;
```
Comment (marek-safar) [inline src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs]: What about releasing the reference on explicit Dispose?

File: src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs [hunk 2]
```diff
@@ -29,6 +29,16 @@ internal class RuntimeAssembly : Assembly
 
         #endregion
 
+        private class ManifestResourceStream : UnmanagedMemoryStream
```
Comment (marek-safar) [inline src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs]: nit: should be sealed

File: src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs (hunk 1 above)
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs]: Explicit Dispose does not guarantee that nobody is accessing the memory anymore

Comment (marek-safar) [reply src/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs]: fair enough, it'd have to invalidate the object state (maybe v2)

**Comment by janvorli** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test please

**Pull request merged**